From 6dc6f13074f4dd14fad4af6b386d53f044eb30c6 Mon Sep 17 00:00:00 2001
From: mengkanglai <mengkanglai2@h-partners.com>
Date: Mon, 15 Dec 2025 17:24:39 +0800
Subject: [PATCH 930/960] downstream: uvb: introduce and export
 odf_get_fdt_ubiostbl

This patch introduces and exports odf_get_fdt_ubiostbl() to retrieve
ubiostbl information from FDT for the uvb external module to use.

Signed-off-by: mengkanglai <mengkanglai2@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 config.aarch64                         |  1 +
 drivers/firmware/Kconfig               |  1 +
 drivers/firmware/Makefile              |  1 +
 drivers/firmware/uvb/Kconfig           | 14 +++++++++++
 drivers/firmware/uvb/Makefile          |  5 ++++
 drivers/firmware/uvb/odf/odf_get_fdt.c | 34 ++++++++++++++++++++++++++
 6 files changed, 56 insertions(+)
 create mode 100644 drivers/firmware/uvb/Kconfig
 create mode 100644 drivers/firmware/uvb/Makefile
 create mode 100644 drivers/firmware/uvb/odf/odf_get_fdt.c

diff --git a/config.aarch64 b/config.aarch64
index 8ee2079af487..f5f1f1e01eb7 100644
--- a/config.aarch64
+++ b/config.aarch64
@@ -11016,5 +11016,6 @@ CONFIG_UB_UBUS=y
 # UB Bus Core Driver
 CONFIG_UB_UBUS_USI=y
 CONFIG_ARM_GIC_V3_ITS_UBUS=y
+CONFIG_UDFI=y
 
 # end of unified bus
diff --git a/drivers/firmware/Kconfig b/drivers/firmware/Kconfig
index 8ecffdce94b1..512d3d92981b 100644
--- a/drivers/firmware/Kconfig
+++ b/drivers/firmware/Kconfig
@@ -314,5 +314,6 @@ source "drivers/firmware/psci/Kconfig"
 source "drivers/firmware/smccc/Kconfig"
 source "drivers/firmware/tegra/Kconfig"
 source "drivers/firmware/xilinx/Kconfig"
+source "drivers/firmware/uvb/Kconfig"
 
 endmenu
diff --git a/drivers/firmware/Makefile b/drivers/firmware/Makefile
index 28fcddcd688f..defec11828e8 100644
--- a/drivers/firmware/Makefile
+++ b/drivers/firmware/Makefile
@@ -38,3 +38,4 @@ obj-y				+= psci/
 obj-y				+= smccc/
 obj-y				+= tegra/
 obj-y				+= xilinx/
+obj-$(CONFIG_UDFI)		+= uvb/
diff --git a/drivers/firmware/uvb/Kconfig b/drivers/firmware/uvb/Kconfig
new file mode 100644
index 000000000000..1fc33d178047
--- /dev/null
+++ b/drivers/firmware/uvb/Kconfig
@@ -0,0 +1,14 @@
+config UDFI
+	bool "UDFI Drivers"
+	depends on ARM64
+	default n
+	help
+	  UBIOS Distributed Firmware Interface (UDFI) support for kernel
+	  requires a UBIOS platform. UDFI provides communication channels
+	  among OS, BIOS and other firmware:
+	  (1) Call ID Service (CIS), by which OS sends a call to BIOS, can be used
+	  to operate specific hardware, read/write BIOS information or call BIOS
+	  functions , e.g., read/write RTC, modify boot options, etc.
+	  (2) Notify ID Information (NII), by which OS receives notification from
+	  BIOS, provides notification mechanism from BIOS to OS. This is useful
+	  when events detected and BIOS needs to inform OS, e.g., RAS events.
diff --git a/drivers/firmware/uvb/Makefile b/drivers/firmware/uvb/Makefile
new file mode 100644
index 000000000000..ff5f558378dc
--- /dev/null
+++ b/drivers/firmware/uvb/Makefile
@@ -0,0 +1,5 @@
+# Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
+# Create      : 2025-04-18
+# Description : cis odf Makefile
+
+obj-$(CONFIG_UDFI) = odf/odf_get_fdt.o
diff --git a/drivers/firmware/uvb/odf/odf_get_fdt.c b/drivers/firmware/uvb/odf/odf_get_fdt.c
new file mode 100644
index 000000000000..b683955e657c
--- /dev/null
+++ b/drivers/firmware/uvb/odf/odf_get_fdt.c
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
+ * Description: ODF get fdt info
+ * Author: mengkanglai
+ * Create: 2025-04-18
+ */
+#include <asm/unaligned.h>
+#include <linux/of_fdt.h>
+#include <linux/libfdt.h>
+#include <linux/byteorder/generic.h>
+#include <linux/printk.h>
+
+int odf_get_fdt_ubiostbl(u64 *phys_addr, char *tbl)
+{
+	int node, len;
+	const void *prop;
+
+	node = fdt_path_offset(initial_boot_params, "/chosen");
+	if (node < 0) {
+		pr_err("failed to get device tree chosen node\n");
+		return -EINVAL;
+	}
+	prop = fdt_getprop(initial_boot_params, node, tbl, &len);
+	if (!prop) {
+		pr_err("failed to get property\n");
+		return -EINVAL;
+	}
+	*phys_addr = (len == 4) ? (u64)be32_to_cpup((const u32 *)prop) :
+			get_unaligned_be64(prop);
+
+	return 0;
+}
+EXPORT_SYMBOL(odf_get_fdt_ubiostbl);
-- 
2.43.0

